src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs

```diff
@@ -1007,6 +1006,91 @@ RemainingTestsForCase makeNext(ArrayBuilder<BoundDagTest> remainingTests)
             }
         }
 
+        /// <summary>
+        /// Determine what we can learn from one successful runtime type test about another planned
+        /// runtime type test for the purpose of building the decision tree.
+        /// We accomodate a special behavior of the runtime here, which does not match the language rules.
+        /// A value of type `int[]` is an "instanceof" (i.e. result of the `isinst` instruction) the type
+        /// `uint[]` and vice versa.  It is similarly so for every pair of same-sized numeric types, and
+        /// arrays of enums are considered to be their underlying type.  We need the dag construction to
+        /// recognize this runtime behavior, so we pretend that matching one of them gives no information
+        /// on whether the other will be matched.  That isn't quite correct (nothing reasonable we do
+        /// could be), but it comes closest to preserving the existing C#7 behavior without undersirable
+        /// side-effects, and permits the code-gen strategy to preserve the dynamic semantic equivalence
+        /// of a switch (on the one hand) and a series of if-then-else statements (on the other).
+        /// See, for example, https://github.com/dotnet/roslyn/issues/35661
+        /// </summary>
+        bool? ExpressionOfTypeMatchesPatternTypeForLearningFromSuccessfulTypeTest(TypeSymbol expressionType, TypeSymbol patternType, ref HashSet<DiagnosticInfo> useSiteDiagnostics)
+        {
+            bool? result = Binder.ExpressionOfTypeMatchesPatternType(_conversions, expressionType, patternType, ref useSiteDiagnostics, out Conversion conversion);
+            return (!conversion.Exists && isRuntimeSimilar(expressionType, patternType))
+                ? null // runtime and compile-time test behavior differ. Pretend we don't know what happens.
+                : result;
+
+            static bool isRuntimeSimilar(TypeSymbol expressionType, TypeSymbol patternType)
+            {
+                while (expressionType is ArrayTypeSymbol { ElementType: var e1, IsSZArray: var sz1, Rank: var r1 } &&
+                       patternType is ArrayTypeSymbol { ElementType: var e2, IsSZArray: var sz2, Rank: var r2 } &&
+                       sz1 == sz2 && r1 == r2)
+                {
+                    if (e1 is NamedTypeSymbol { TypeKind: TypeKind.Enum, EnumUnderlyingType: var t1 })
+                        e1 = t1;
+
+                    if (e2 is NamedTypeSymbol { TypeKind: TypeKind.Enum, EnumUnderlyingType: var t2 })
+                        e2 = t2;
```

Comment:
There is a `TypeSymbolExtensions.EnumUnderlyingType(this TypeSymbol type)` helper method.